nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py

```diff
@@ -244,26 +258,26 @@ def main() -> None:
         subprocess.check_call(["@systemd@/bin/bootctl", "--path=@efiSysMountPoint@"] + flags + ["install"])
     else:
         # Update bootloader to latest if needed
-        systemd_version = subprocess.check_output(["@systemd@/bin/bootctl", "--version"], universal_newlines=True).split()[2]
-        sdboot_status = subprocess.check_output(["@systemd@/bin/bootctl", "--path=@efiSysMountPoint@", "status"], universal_newlines=True)
+        systemd = subprocess.check_output(["@systemd@/bin/bootctl", "--version"], universal_newlines=True).split()[2]
+        sdboot = subprocess.check_output(["@systemd@/bin/bootctl", "--path=@efiSysMountPoint@", "status"], universal_newlines=True)
 
         # See status_binaries() in systemd bootctl.c for code which generates this
-        m = re.search("^\W+File:.*/EFI/(BOOT|systemd)/.*\.efi \(systemd-boot ([\d.]+[^)]*)\)$",
-                      sdboot_status, re.IGNORECASE | re.MULTILINE)
-
-        needs_install = False
-
-        if m is None:
-            print("could not find any previously installed systemd-boot, installing.")
-            # Let systemd-boot attempt an installation if a previous one wasn't found
-            needs_install = True
-        else:
-            sdboot_version = f'({m.group(2)})'
-            if systemd_version != sdboot_version:
-                print("updating systemd-boot from %s to %s" % (sdboot_version, systemd_version))
-                needs_install = True
-
-        if needs_install:
+        sdboot_version_match = re.search("^\W+File:.*/EFI/(BOOT|systemd)/.*\.efi \(systemd-boot ([\d.]+[^)]*)\)$",
+                      sdboot, re.IGNORECASE | re.MULTILINE)
+
+        systemd_version_match = re.search("^\((.*)\)$", systemd)
+
+        if sdboot_version_match is None:
+            raise Exception("could not find any previously installed systemd-boot")
+
+        if systemd_version_match is None:
+            raise Exception("could not determine systemd-boot version")
+
+        sdboot_version = sdboot_version_match.group(2)
+        systemd_version = systemd_version_match.group(1)
+
+        if should_update(sdboot_version, systemd_version):
+            print("updating systemd-boot from %s to %s" % (sdboot_version, systemd_version))
             subprocess.check_call(["@systemd@/bin/bootctl", "--path=@efiSysMountPoint@", "update"])
```

Comment:
Should probably print out some sort of "Not downgrading systemd-boot" message to let users know that their configuration isn't getting fully applied. Otherwise all seems reasonable to me :+1:

Reply:
I agree, I'll update the commit tomorrow.

Reply:
I added an `else` case to inform the user the script chose to not update `systemd-boot`. I struggled a bit to find a wording that includes both, the installed and available version which is not too awkward. I hope it's ok this way.